tools/bigquery-hive-external-table-loader/src/main/scala/com/google/cloud/bqhiveloader/SparkJobs.scala

```diff
@@ -0,0 +1,399 @@
+    val metaStore = {
+      config.hiveMetastoreType match {
+        case "jdbc" =>
+          JDBCMetaStore(config.hiveJdbcUrl, spark)
+        case "sql" =>
+          SparkSQLMetaStore(spark)
+        case "external" =>
+          ExternalCatalogMetaStore(spark)
+        case x =>
+          throw new IllegalArgumentException(s"unsupported metastore type '$x'")
+      }
+    }
+
+    for {
+      keytab <- config.krbKeyTab
+      principal <- config.krbPrincipal
+    } yield {
+      Kerberos.configureJaas(keytab, principal)
+    }
+
+    runWithMetaStore(config, metaStore, spark)
+  }
+
+  /** Reads partitions from MetaStore and launches Spark Job
+    *
+    * @param config
+    * @param metaStore
+    * @param spark
+    */
+  def runWithMetaStore(config: Config, metaStore: MetaStore, spark: SparkSession): Unit = {
+    val table = metaStore.getTable(config.hiveDbName, config.hiveTableName)
+    val partitions: Seq[Partition] =
+      if (config.partitioned && table.partitionColumnNames.nonEmpty) {
+        metaStore.filterPartitions(db = config.hiveDbName,
+          table = config.hiveTableName,
+          filterExpression = config.partFilters
+        ) match {
+          case parts if parts.nonEmpty =>
+            logger.info(s"Selected ${parts.length} partitions for table '${config.hiveDbName}.${config.hiveTableName}':\n\t${parts.map(_.location).mkString("\n\t")}")
+            parts
+          case _ =>
+            throw new RuntimeException(s"No partitions found with filter expression '${config.partFilters}'")
+        }
+      } else {
+        table.location match {
+          case Some(location) =>
+            logger.info(s"Loading '${config.hiveDbName}.${config.hiveTableName}' as non-partitioned table from $location")
+            Seq(Partition(Seq.empty, location))
+          case _ =>
+            throw new RuntimeException("Location not found in table description for non-partitioned table")
+        }
+      }
+    val sc = spark.sparkContext
+    sc.runJob(rdd = sc.makeRDD(Seq(config), numSlices = 1),
+              func = loadPartitionsJob(table, partitions))
+  }
+
+  /** Spark Job
+    *
+    * @param table
+    * @param partitions
+    * @return
+    */
+  def loadPartitionsJob(table: TableMetadata, partitions: Seq[Partition]): Iterator[Config] => Unit =
+    (it: Iterator[Config]) => loadPartitions(it.next, table, partitions)
+
+  def readLocalOrSparkFile(maybePath: scala.Option[String]): scala.Option[Array[Byte]] = {
+    val direct = maybePath
+      .map(Paths.get(_))
+      .filter{_.toFile.exists()}
+
+    val fromSparkFiles = maybePath
+      .map(SparkFiles.get)
+      .map(Paths.get(_))
+      .filter{_.toFile.exists}
+
+    direct.orElse(fromSparkFiles)
+      .map(Files.readAllBytes)
+      .filter(_.nonEmpty)
+  }
+
+  /** Spark Job to create external table and create BigQuery load job
+    * The data is not touched, this is launched as a remote job to use
```

Comment:
....create BigQuery load job to load the data is not touched. The Spark Job is launched.....

Reply:
Thanks, added more detail here.